File: api/src/main/java/com/openlattice/graph/GraphApi.java
/*
 * Copyright (C) 2018. OpenLattice, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * You can contact the owner of the copyright at support@openlattice.com
 */

package com.openlattice.graph;

import com.google.common.collect.ListMultimap;
import com.google.common.collect.SetMultimap;
import com.openlattice.graph.query.GraphQuery;
import com.openlattice.graph.query.GraphQueryState;
import com.openlattice.graph.query.GraphQueryState.Option;
import java.util.Set;
import java.util.UUID;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;

/**
 * Used for performing graph queries on the backend.
 */
public interface GraphApi {
    String CONTROLLER = "/graph";
    String ID      = "id";
    String ID_PATH = "/{" + ID + "}";
    //@formatter:on
    String QUERY  = "/query";
    String RESULT = "/result";
    //@formatter:off
    String SERVICE    = "/datastore";
    String BASE       = SERVICE + CONTROLLER;

    @POST( BASE + QUERY )
    GraphQueryState submit( SimpleGraphQuery query );

    @POST( BASE + QUERY + ID_PATH )
    GraphQueryState getQueryState( @Path( ID ) UUID queryId, @Body Set<Option> options );

    /**
     * Retrieves the graph query state with any additional options. Equivalent to {@code getQueryState(queryId, Collections.EMPTY_SET) }
     */
    @GET( BASE + QUERY + ID_PATH )
    GraphQueryState getQueryState( @Path( ID ) UUID queryId );

    @GET( BASE + QUERY + ID_PATH + RESULT )
    SubGraph getResults( UUID queryId );

    /**
     *
     * @param ops
     * @return
     */
    ListMultimap<UUID, SetMultimap<UUID, SetMultimap<UUID, Object>>> graphQuery( GraphQuery ops ); // Entity Key Id -> Property Type Id -> Entity Set Id
    //That is a list of entities along with properties annotated with which entity set they came from.

}

File: api/src/main/kotlin/com/openlattice/graph/GraphApi.kt
/*
 * Copyright (C) 2019. OpenLattice, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * You can contact the owner of the copyright at support@openlattice.com
 *
 *
 */

package com.openlattice.graph

import retrofit2.http.Body
import retrofit2.http.POST
import retrofit2.http.Path
import java.util.*


//@formatter:off
const val SERVICE = "/datastore"
const val CONTROLLER = "/graph"
const val BASE = SERVICE + CONTROLLER
//@formatter:on

const val ENTITY_SET_ID = "entitySetId"
const val ENTITY_SET_ID_PATH = "/{$ENTITY_SET_ID}"
const val ID = "id"
const val ID_PATH = "/{$ID}"
const val NEIGHBORS = "/neighbors"
const val PAGE = "/page"
const val QUERY = "/query"

/**
 *
 * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
 */
interface GraphApi {

    /**
     * Executes a search for all neighbors of multiple entities of the same entity set that are connected by an
     * association
     *
     * @param entitySetId The entity set id of the entities
     * @param query      Constraints and filters on the neighborhood to return
     * @return A map from each entity id to a list of objects containing information about the neighbors and
     * associations of that entity
     */
    @POST(BASE + NEIGHBORS + ENTITY_SET_ID_PATH)
    fun neighborhoodQuery(
            @Path(ENTITY_SET_ID) entitySetId: UUID,
            @Body query: NeighborhoodQuery
    ): Neighborhood

    /**
     * Loads a page of neighbors for the given entity set as defined by the pagedNeighborRequest.
     *
     * @param entitySetId The id of the entity set to load neighbors for
     * @param pagedNeighborRequest An object defining which entities to load neighbors for, what entity sets to load
     * from, how many neighbors to return, and a paging token indicating where to begin the page of results.
     *
     * @return A [NeighborPage] object, containing up to the requested number of neighbors, and also a bookmark to be
     * used as a paging token for subsequent requests.
     *
     */
    @POST(BASE + NEIGHBORS + ENTITY_SET_ID_PATH + PAGE)
    fun getPageOfNeighbors(
            @Path(ENTITY_SET_ID) entitySetId: UUID,
            @Body pagedNeighborRequest: PagedNeighborRequest
    ): NeighborPage
}